src/node_buffer.cc

```diff
@@ -1051,6 +1054,118 @@ static void EncodeUtf8String(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+static void EncodeInto(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  Local<Context> context = env->context();
+  CHECK_GE(args.Length(), 2);
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsUint8Array());
+
+  size_t read = 0;
+  size_t written = 0;
+
+  Utf8Value src(isolate, args[0]);
+  const char* p = *src;
+
+  Local<Uint8Array> dest = args[1].As<Uint8Array>();
+  Local<ArrayBuffer> buf = dest->Buffer();
+  char* write_result =
+      static_cast<char*>(buf->GetContents().Data()) + dest->ByteOffset();
+  size_t dest_length = dest->ByteLength();
+
+  for (size_t i = 0; i < src.length(); ) {
+    char c = p[i];
+    uint32_t code = 0;
+
+    if ((c & 0x80) == 0) {
+      code = c;
+      i += 1;
+    } else if ((c & 0xE0) == 0xC0) {
+      code = (p[i] & 0x1F) << 6;
+      code |= (p[i+1] & 0x3F);
+      i += 2;
+    } else if ((c & 0xF0) == 0xE0) {
+      code = (p[i] & 0xF) << 12;
+      code |= (p[i+1] & 0x3F) << 6;
+      code |= (p[i+2] & 0x3F);
+      i += 3;
+    } else if ((c & 0xF8) == 0xF0) {
+      code = (p[i] & 0x7) << 18;
+      code |= (p[i+1] & 0x3F) << 12;
+      code |= (p[i+2] & 0x3F) << 6;
+      code |= (p[i+3] & 0x3F);
+      i += 4;
+    } else if ((c & 0xFC) == 0xF8) {
+      code = (p[i] & 0x3) << 24;
+      code |= (p[i] & 0x3F) << 18;
+      code |= (p[i] & 0x3F) << 12;
+      code |= (p[i] & 0x3F) << 6;
+      code |= (p[i] & 0x3F);
+      i += 5;
+    } else if ((c & 0xFE) == 0xFC) {
+      code = (p[i] & 0x1) << 30;
+      code |= (p[i] & 0x3F) << 24;
+      code |= (p[i] & 0x3F) << 18;
+      code |= (p[i] & 0x3F) << 12;
+      code |= (p[i] & 0x3F) << 6;
+      code |= (p[i] & 0x3F);
+      i += 6;
+    }
+
+    if (code <= 0x7F) {
+      if (dest_length < 1) break;
+
+      *write_result++ = static_cast<char>(code);
```

Comment:
We’re still essentially decoding and re-encoding the individual code points here… in the end, the target buffer and `*src` will have the same contents up to a given index, right? So wouldn’t it be enough to figure out how many bytes we want to copy in total, and then to do that?

Reply:
@addaleax I can not agree more with you. Return value `read` required to be the read Unicode code units numbers, the `src` is encoded UTF8 in order to the calculate the Unicode code units numbers I translate the UTF8  character into Unicode.